episodes/12-fire-forget-tasks.md

```diff
@@ -0,0 +1,392 @@
+coforall index in iterand
+{instructions}
+```
+
+This will start a new task, for each iteration. Each tasks will then perform all the instructions inside the
+curly brackets. Each task will have a copy of the variable **_index_** with the corresponding value yielded by
+the iterand.  This index allows us to _customise_ the set of instructions for each particular task.
+
+```chpl
+var x=1;
+config var numoftasks=2;
+
+writeln("This is the main task: x = ",x);
+
+coforall taskid in 1..numoftasks do
+{
+  var c=taskid+1;
+  writeln("this is task ",taskid,": x + ",taskid," = ",x+taskid,". My value of c is: ",c);
+}
+
+writeln("this message won't appear until all tasks are done...");
+```
+
+```bash
+> > chpl coforall_example.chpl -o coforall_example
+> > ./coforall_example --numoftasks=5
+```
+
+```output
+This is the main task: x = 1
+this is task 5: x + 5 = 6. My value of c is: 6
+this is task 2: x + 2 = 3. My value of c is: 3
+this is task 4: x + 4 = 5. My value of c is: 5
+this is task 3: x + 3 = 4. My value of c is: 4
+this is task 1: x + 1 = 2. My value of c is: 2
+this message won't appear until all tasks are done...
+```
+
+Notice how we are able to customise the instructions inside the coforall, to give different results depending
+on the task that is executing them. Also, notice how, once again, the variables declared outside the coforall
+can be read by all tasks, while the variables declared inside, are available only to the particular task.
+
+::::::::::::::::::::::::::::::::::::: challenge
+
+## Challenge 1: Can you do it?
+
+Would it be possible to print all the messages in the right order? Modify the code in the last example as
+required.
+
+Hint: you can use an array of strings declared in the main task, where all the concurrent tasks could write
+their messages in the corresponding position. Then, at the end, have the main task printing all elements of
+the array in order.
+
+:::::::::::::::::::::::: solution
+
+The following code is a possible solution:
+
+```chpl
+var x=1;
+config var numoftasks=2;
+var messages: [1..numoftasks] string;
+
+writeln("This is the main task: x = ",x);
+
+coforall taskid in 1..numoftasks do
+{
+   var c=taskid+1;
+   var s="this is task "+taskid+": x + "+taskid+" = "+(x+taskid)+". My value of c is: "+c;
+   messages[taskid]=s;
+}
+
+for i in 1..numoftasks do writeln(messages[i]);
+writeln("this message won't appear until all tasks are done...");
+```
+
+```bash
+chpl exercise_coforall.chpl -o exercise_coforall
+./exercise_coforall --numoftasks=5
+```
+
+```output
+This is the main task: x = 1
+this is task 1: x + 1 = 2. My value of c is: 2
+this is task 2: x + 2 = 3. My value of c is: 3
+this is task 3: x + 3 = 4. My value of c is: 4
+this is task 4: x + 4 = 5. My value of c is: 5
+this is task 5: x + 5 = 6. My value of c is: 6
+this message won't appear until all tasks are done...
+```
+
+Note that `+` is a **_polymorphic_** operand in Chapel. In this case it concatenates `strings` with `integers`
```

Comment:
Define polymorphic.

Reply:
I updated the code in this example, as Chapel 2.0 no longer can add strings and integers, and you first need to convert an integer to a string. I am not sure if *polymorphic* is the right term here at all (added by someone else), so I rewrote the sentence without it.